### PR TITLE
feature/se-6-openaire

### LINKED ISSUE #17: feature/se-6-openAIRE
Create a new component that will use autocomplete.

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|-----:|-----:|-------:|
| ETA                  |  9.3   |    0 |     0 |      0 |        0 |
| Developing      |  3+8+4+7+4+6+8+4+8+8+8+5+4+6+6+5+4+7+1+1+1.5  |    0 |    0 |      0 |         0 |
| Review             |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -   |   -     |        0 |
## Problem description
Show funding suggestions for EU projects - openAIRE and non-EU projects. Suggestions are showed in the complex input field.

- [x] For EU projects load suggestions from the openAIRE endpoint.
- [x] For non-EU projects load suggestions from the metadata suggestions endpoint.
- [x] Create tests

**!!!** This pull request contains pull request for _complex-input-types_ and _autocomplete_ **!!!**
**complex-input-types:** https://github.com/dataquest-dev/dspace-angular/pull/28
**autocomplete:** https://github.com/dataquest-dev/dspace-angular/pull/30
### Reported issues
https://github.com/dataquest-dev/dspace-angular/issues/54

### 1. Github action problem
Tests run locally but it fail in the github action. **Reason**: In the github action tests runs with code from dtq-dev.

### 2. Integration test problem
**Problem:** Integration tests for openAIRE was failing in different time on different places. Sometimes they pass sometimes they falls.
**Solution:** If the tests are run multiple times - they pass. I changed `retries` property to 6.

### CODE REVIEW

File: cypress/integration/submission-ui.spec.ts [hunk 1]
```diff
@@ -1,5 +1,10 @@
-const password = 'admin';
-const email = 'test@test.edu';
+/**
+ * This IT will be never be pushed to the upstream because clicking testing DOM elements is antipattern because
+ * the tests on other machines could be failed.
```
Comment (vidiecan) [inline cypress/integration/submission-ui.spec.ts]:
```suggestion
 * the tests on other machines could fail.
```

Comment (vidiecan) [reply cypress/integration/submission-ui.spec.ts]: can we prefix if with clarin-dspace or similar? So we can distinguish them easier

File: cypress/integration/submission-ui.spec.ts [hunk 2]
```diff
@@ -1,5 +1,10 @@
-const password = 'admin';
-const email = 'test@test.edu';
+/**
+ * This IT will be never be pushed to the upstream because clicking testing DOM elements is antipattern because
```
Comment (vidiecan) [inline cypress/integration/submission-ui.spec.ts:2]: how is vanilla dspace doing it? No DOM?

File: src/app/core/data/metadata-value-data.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import {Observable, of as observableOf} from 'rxjs';
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.spec.ts]: what is this test doing?
ADD this description into every test

File: src/app/core/data/metadata-value-data.service.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  ) {
+    super();
+  }
+
+  /**
+   * Retrieve the MetadataValue object inside Vocabulary object body
+   */
+  findByMetadataNameAndByValue(metadataName, term): Observable<PaginatedList<MetadataValue>> {
+      const metadataFields = metadataName.split('.');
+      const optionParams = Object.assign(new FindListOptions(), {}, {
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.ts:70]: what if the metadataName does not have 2 dots

File: src/app/core/data/metadata-value-data.service.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        searchParams: [
+          new RequestParam('schema', metadataFields[0]),
+          new RequestParam('element', metadataFields[1]),
+          new RequestParam('qualifier', metadataFields[2]),
+          new RequestParam('searchValue', term)
+        ]
+      });
+      const remoteData$ = this.searchBy('byValue', optionParams);
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.ts:78]: why `$`? is it observable?

File: src/app/core/data/metadata-value-data.service.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+      return remoteData$.pipe(
+        getFirstSucceededRemoteDataPayload(),
+        map((list: PaginatedList<MetadataValue>) => {
+          const vocabularyEntryList: VocabularyEntry[] = [];
+          for (const metadataValue of list.page) {
+            const voc: VocabularyEntry = new VocabularyEntry();
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.ts:85]: why this notation?

File: src/app/core/metadata/metadata-value.model.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
+import {link, typedObject} from '../cache/builders/build-decorators';
```
Comment (vidiecan) [inline src/app/core/metadata/metadata-value.model.ts]: this class (or similar) was not there before?

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 8]
```diff
@@ -0,0 +1,296 @@
+import {ChangeDetectorRef, Component, EventEmitter, Input, OnInit, Output, TemplateRef, ViewChild} from '@angular/core';
+import {FormGroup} from '@angular/forms';
+import {DynamicTagModel} from '../tag/dynamic-tag.model';
+import {NgbTypeahead, NgbTypeaheadSelectItemEvent} from '@ng-bootstrap/ng-bootstrap';
+import {Observable, of as observableOf} from 'rxjs';
+import {PageInfo} from '../../../../../../core/shared/page-info.model';
+import {VocabularyService} from '../../../../../../core/submission/vocabularies/vocabulary.service';
+import {DynamicFormLayoutService, DynamicFormValidationService} from '@ng-dynamic-forms/core';
+import {catchError, debounceTime, distinctUntilChanged, map, merge, switchMap, tap} from 'rxjs/operators';
+import {buildPaginatedList, PaginatedList} from '../../../../../../core/data/paginated-list.model';
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: with so many ../ .. it is like ASCII art

╱╱▏╱▏╱▏╱╱╱╱╱╱▏╱╱╱╱╱▏
▉╱▉╱▉╱▏▉▉▉▉▉╱┈▉▉▉▉╱
▉╱▉╱▉╱▏┈┈▉╱▏┈┈▉╱╱╱╱▏
▉╱▉╱▉╱▏┈┈▉╱▏┈┈▉▉▉▉╱
▉▉▉▉▉╱┈┈┈▉╱┈┈┈▉╱

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    this.cdr.detectChanges();
+  }
+
+  updateModel(updateValue) {
+    let newValue = updateValue.display;
+    if (this.isSponsorInputType) {
+      if (updateValue instanceof VocabularyEntry) {
+        newValue = AUTOCOMPLETE_COMPLEX_PREFIX + SEPARATOR + updateValue.value;
+      } else {
+        // special autocomplete sponsor input
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: if it is special, this should be guarded by if (...)

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+  suggestionFormatter = (suggestion: TemplateRef<any>) => {
+    if (this.isSponsorInputType) {
+      let formattedValue = '';
+      // normal MetadataValue suggestion
+      if (suggestion instanceof VocabularyEntry) {
+        formattedValue = suggestion.value;
+        if (formattedValue.startsWith(AUTOCOMPLETE_COMPLEX_PREFIX)) {
+          formattedValue = DynamicAutocompleteService.removeAutocompletePrefix(suggestion);
+        }
+        const complexInputList = formattedValue.split(SEPARATOR);
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: make sure it has 2 parts

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        return DynamicAutocompleteService.pretifyFundingSuggestion(complexInputList[1], complexInputList[2]);
+      } else if (suggestion instanceof ExternalSourceEntry) {
+        // suggestion from the openAIRE
+        const fundingProjectCode = this.getProjectCodeFromId(suggestion.id);
+        const fundingName = suggestion.metadata['project.funder.name'][0].value;
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: can it b eempty?

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    return suggestion.display;
+  }
+
+  /**
+   * Converts a text values stream from the `<input>` element to the array stream of the items
+   * and display them in the typeahead popup.
+   */
+  search = (text$: Observable<string>) =>
+    text$.pipe(
+      debounceTime(300),
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: ?

Comment (vidiecan) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: why 300

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+      switchMap((term) => {
+        // min 3 characters
+        if (term === '' || term.length < this.model.minChars) {
+          return observableOf({ list: [] });
+        } else {
+          let response: Observable<PaginatedList<ExternalSourceEntry | MetadataValue>>;
+          if (this.isSponsorInputType) {
+            // if openAIRE
+            // @ts-ignore
+            const fundingType = this.model.parent.group[0].value;
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: 0

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+  /**
+   * Only for the local.sponsor complex input type
+   * The external funding is composed as one complex input field
+   * @param updateValue external funding from the openAIRE
+   */
+  composeSponsorInput(updateValue) {
+    // set prefix to distinguish composed complex input in the complex.model.ts - get method
+    let newValue = AUTOCOMPLETE_COMPLEX_PREFIX + SEPARATOR;
+    let fundingType = 'N/A';
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: localisation? global settings?

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    let fundingProjectCode = '';
+
+    if (updateValue.id.startsWith(EU_PROJECT_PREFIX)) {
+      fundingType = 'EU';
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: either it is a generic component or very specific

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+   * @param id
+   * @return Funder/FundingProgram/ProjectID
+   */
+  getProjectCodeFromId(id) {
+    let funder = '';
+    let fundingProgram = '';
+    let projectID = '';
+
+    id = id.substr(id.indexOf('grantAgreement'));
+    id = id.substr(id.indexOf('/') + 1);
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: can't you use regex? group matching?

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+  constructor(config: DynamicAutocompleteModelConfig, layout?: DynamicFormControlLayout) {
+
+    super(config, layout);
+
+    if (isEmpty(this.vocabularyOptions)) {
+      this.vocabularyOptions = new VocabularyOptions('none');
+    }
+    this.autoComplete = AUTOCOMPLETE_OFF;
+    this.minChars = config.minChars || 3;
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts]: it is like magic

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.service.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
+import {AUTOCOMPLETE_COMPLEX_PREFIX} from './dynamic-autocomplete.model';
+import {SEPARATOR} from '../ds-dynamic-complex.model';
+
+export class DynamicAutocompleteService {
+  static removeAutocompletePrefix(formValue) {
+    return formValue.value.replace(AUTOCOMPLETE_COMPLEX_PREFIX + SEPARATOR, '');
+  }
+
+  static pretifyFundingSuggestion(fundingProjectCode, fundingName) {
+    return 'Funding code: '.bold() + fundingProjectCode + '<br> Project name: '.bold() + fundingName;
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.service.ts]: loc

File: src/app/shared/form/builder/parsers/autocomplete-field-parser.ts [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+import {
+  DynamicAutocompleteModel,
+  DynamicAutocompleteModelConfig
+} from '../ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model';
+
+export class AutocompleteFieldParser extends FieldParser {
+
+  public modelFactory(fieldValue?: FormFieldMetadataValueObject | any, label?: boolean): any {
+    const autocompleteModelConfig: DynamicAutocompleteModelConfig = this.initModel(null, label);
+    if (this.configData.selectableMetadata[0].controlledVocabulary
```
Comment (vidiecan) [inline src/app/shared/form/builder/parsers/autocomplete-field-parser.ts]: [0]

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: cypress/integration/submission-ui.spec.ts (hunk 2 above)
Comment (milanmajchrak) [reply cypress/integration/submission-ui.spec.ts:2]: vanilla dspace is testing DOM elements but it has very simple integration tests - it check only visibility or main simple functions of the DOM elements. Our tests are more difficult.

File: cypress/integration/submission-ui.spec.ts (hunk 1 above)
Comment (milanmajchrak) [reply cypress/integration/submission-ui.spec.ts]: yes we can - done

File: src/app/core/data/metadata-value-data.service.ts (hunk 4 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts:70]: Fixed

File: src/app/core/data/metadata-value-data.service.ts (hunk 5 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts:78]: Yes

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.service.ts (hunk 18 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.service.ts]: done

File: src/app/shared/form/builder/parsers/autocomplete-field-parser.ts (hunk 19 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/parsers/autocomplete-field-parser.ts]: fixed

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts (hunk 17 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts]: done

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 16 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: fixed - used regex

Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: used regex

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 15 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: I created DsDynamicSponsorAutocomplete component which extends DsDynamicAutocomplete

File: src/app/core/metadata/metadata-value.model.ts (hunk 7 above)
Comment (milanmajchrak) [reply src/app/core/metadata/metadata-value.model.ts]: There are similar metadataValue classes but no one is capable for our issue.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 8 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: Yes, but imports like that are through whole dspace-angular. example: https://github.com/dataquest-dev/dspace-angular/blob/dtq-dev/src/app/item-page/edit-item-page/item-status/item-status.component.ts
These imports are generated automatically.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 9 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: refactored - special entry is ExternalSourceEntry. I added a check for that.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 10 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: this code no longer exists

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 11 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: added a check there according to https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 12 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: I don't know I copied it from the _dynamic-tag.component_ (DSpace class). But I think 300ms as debounce time (time between typing a characters) is good option.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 13 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: Added a check

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts (hunk 14 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts]: fixed - value are loaded from the en.json

File: src/app/core/data/metadata-value-data.service.spec.ts (hunk 3 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.spec.ts]: doc added

File: src/app/core/data/metadata-value-data.service.ts (hunk 6 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts:85]: for replaced by foreach

File: cypress/integration/submission-ui.spec.ts [hunk 20]
```diff
@@ -117,23 +174,77 @@ describe('Create a new submission', () => {
     cy.visit('/');
     sideBarMenu.clickOnNewButton();
     sideBarMenu.clickOnNewItemButton();
+    createItemProcess.typeCollectionName();
     createItemProcess.selectCollection();
   });
 
+  // @TODO uncomment this tests when the openAIRE will be merged
```
Comment (mrovnanik) [inline cypress/integration/submission-ui.spec.ts]: uncomment

**Review by mrovnanik — COMMENTED** [review]
remove comments

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts (hunk 17 above)
Comment (mrovnanik) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts]: Add a comment why 3 is a good constant. The idea is to have some kind of description. In a sense that we think its useful to start searching for matches after the user enters at least 3 chars ....

Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.model.ts]: This conversation is outdated. I added constant there, new code: 
image

**Review by mrovnanik — APPROVED** [review]
tests commented out for now

**Pull request merged**